clippy_lints/src/methods/map_unwrap_or.rs

```diff
@@ -43,19 +45,33 @@ pub(super) fn lint<'a, 'tcx>(
         let map_snippet = snippet(cx, map_args[1].span, "..");
         let unwrap_snippet = snippet(cx, unwrap_args[1].span, "..");
         // lint message
-        // comparing the snippet from source to raw text ("None") below is safe
-        // because we already have checked the type.
-        let arg = if unwrap_snippet == "None" { "None" } else { "a" };
-        let suggest = if unwrap_snippet == "None" {
-            "and_then(f)"
+        let msg = if is_option {
+            // comparing the snippet from source to raw text ("None") below is safe
+            // because we already have checked the type.
+            let (arg, suggest) = if unwrap_snippet == "None" {
+                ("None", "and_then(f)")
+            } else {
+                ("a", "map_or(a, f)")
+            };
+
+            format!(
+                "called `map(f).unwrap_or({})` on an Option value. \
+                 This can be done more directly by calling `{}` instead",
+                arg, suggest
+            )
         } else {
-            "map_or(a, f)"
+            debug_assert!(is_result);
+            "called `map(f).unwrap_or(a)` on a Result value. \
+             This can be done more directly by calling `map_or(a, f)` instead"
```

Comment:
Can you do this by using the function
```rust
span_lint_and_then(
    ..,
    |db| {
         db.note("this can be done more directly by calling `map_or(a, f)` instead");
         if multiline {
               db.span_help(.., format!("replace `map({}).unwrap_or({})` with `{}`", ..);
         }
    },
);
```